The Matthews Correlation Coefficient can be outside the interval [-1, 1] in some rare cases

The Matthews Correlation Coefficient should always lie in the range [-1, 1] based on its definition. However, due to some floating-point issues, hypothesis has found a couple of falsifying examples. This causes the hypothesis test to be flaky, as it doesn't always find a falsifying example. We should add a falsifying example as a test and then fix the issue.

Since this only seems to happen with truly massive numbers (in the quintillion range), I think that we're safe to just clamp the result of `model_metrics.mcc()` before returning. It's unlikely that we will ever have a dataset with 2_070_366_244_862_899 false negatives!

```python
   def test_mcc_is_between_negative_1_and_positive_1_massive_numbers_example() -> None:
        """
        Falsifying example found by hypothesis.
        """
        mcc_score = mcc(0, 0, 51, 2_070_366_244_862_899)
>       assert -1.0 <= mcc_score <= 1.0
E       assert -1.0 <= -1.0000000000000002
```

--- a/hlink/linking/core/model_metrics.py
+++ b/hlink/linking/core/model_metrics.py
@@ -16,6 +16,16 @@ such as cases with division by zero.
 import math
 
 
+def clamp(value: float, minimum: float, maximum: float) -> float:
+    """
+    Clamp the given value, forcing it to be between the minimum and maximum.
+    """
+    if minimum > maximum:
+        raise ValueError("minimum is greater than maximum")
+
+    return max(minimum, min(value, maximum))
+
+
 def f_measure(true_pos: int, false_pos: int, false_neg: int) -> float:
     """
     Compute the F-measure, which is defined as the harmonic mean of precision
@@ -60,7 +70,8 @@ def mcc(true_pos: int, true_neg: int, false_pos: int, false_neg: int) -> float:
         return math.nan
 
     numerator = true_pos * true_neg - false_pos * false_neg
-    return numerator / denominator
+    value = numerator / denominator
+    return clamp(value, minimum=-1.0, maximum=1.0)
 
 
 def precision(true_pos: int, false_pos: int) -> float:

--- a/hlink/tests/core/model_metrics_test.py
+++ b/hlink/tests/core/model_metrics_test.py
@@ -4,14 +4,15 @@
 #   https://github.com/ipums/hlink
 import math
 
-from hypothesis import assume, given
+from hypothesis import assume, example, given
 import hypothesis.strategies as st
 import pytest
 
-from hlink.linking.core.model_metrics import f_measure, mcc, precision, recall
+from hlink.linking.core.model_metrics import clamp, f_measure, mcc, precision, recall
 
 NonNegativeInt = st.integers(min_value=0)
 NegativeInt = st.integers(max_value=-1)
+BoundedFloat = st.floats(allow_infinity=False, allow_nan=False)
 
 
 def test_f_measure_example() -> None:
@@ -77,6 +78,9 @@ def test_mcc_example() -> None:
     true_neg=NonNegativeInt,
     false_pos=NonNegativeInt,
     false_neg=NonNegativeInt,
+)
+@example(true_pos=0, true_neg=0, false_pos=51, false_neg=2_070_366_244_862_899).via(
+    "issue #187"
 )
 def test_mcc_is_between_negative_1_and_positive_1(
     true_pos: int, true_neg: int, false_pos: int, false_neg: int
@@ -167,3 +171,30 @@ def test_recall_no_true_pos_or_false_neg() -> None:
     """
     recall_score = recall(0, 0)
     assert math.isnan(recall_score)
+
+
+def test_clamp_in_between() -> None:
+    assert clamp(15, 10, 20) == 15
+
+
+def test_clamp_less_than_minimum() -> None:
+    assert clamp(1, 5, 10) == 5
+
+
+def test_clamp_greater_than_maximum() -> None:
+    assert clamp(200, 10, 30) == 30
+
+
+@given(x=BoundedFloat, y=BoundedFloat, z=BoundedFloat)
+def test_clamp_lies_within_bounds(x: float, y: float, z: float) -> None:
+    assume(y <= z)
+    assert y <= clamp(x, y, z) <= z
+
+
+@given(x=BoundedFloat, y=BoundedFloat, z=BoundedFloat)
+def test_clamp_error_when_minimum_greater_than_maximum(
+    x: float, y: float, z: float
+) -> None:
+    assume(y > z)
+    with pytest.raises(ValueError, match="minimum is greater than maximum"):
+        clamp(x, y, z)

